std/typecons.d

```diff
@@ -7093,3 +7094,535 @@ unittest
+        }
+    }
+
+    /**
+     * Range primitives.
+     *
+     * `Option` is either a single-element range containing the value it holds
+     * or an _empty range when no value is held.
+     *
+     * Slicing is only provided for mutable types `T`.
+     *
+     * Note: Using front or back _to nullify the held value results in a
+     * $(I non-_empty) null instance. Only popping or assigning _to $(LREF none)
+     * empties the instance.
+     * See_Also: $(MREF std, range)
+     */
+    ref inout(T) front() inout @property { assert(isDefined); return value; }
+    ref inout(T) back() inout @property { assert(isDefined); return value; } /// Ditto
+    bool empty() const @property { return !isDefined; } /// Ditto
+    void popFront() { this = none; } /// Ditto
+    void popBack() { this = none; } /// Ditto
+    static if (isCopyable!T) Option save() @property { return this; } /// Ditto
+    size_t length() const @property { return isDefined; } /// Ditto
+    alias opDollar = length; /// Ditto
+
+    /// Ditto
+    ref inout(T) opIndex(size_t index) inout
+    {
+        version(assert)
+        {
+            import core.exception : RangeError;
+            if (!isDefined || index != 0) throw new RangeError();
+        }
+        return value;
+    }
+
+    static if (isCopyable!T)
+    {
+        /// Ditto
+        Option opSlice(size_t from, size_t to)
+        {
+            version(assert)
+            {
+                import core.exception : RangeError;
+                if (from > to || length < to) throw new RangeError();
+            }
+            return to > from ? this : Option(none);
+        }
+    }
+
+    /**
+     * Assume this instance holds a value and return it.
+     *
+     * Range algorithms are preferable over raw access to the underlying
+     * value when possible. In other cases, using Option and assumeNonEmpty is
+     * preferable to using a raw nullable reference, as the `AssertError` is
+     * guaranteed to be thrown, and unlike a segmentation fault or other
+     * undefined behavior, will be as close to the underlying error as
+     * possible. Further, using assumeNonEmpty formalizes the assumption of
+     * non-null.
+     * Returns: The held value
+     * Throws: `AssertError` if this instance is empty
+     * Note: `opt.assumeNonEmpty = null;` does not empty the instance; to empty
+     * the instance, assign to $(LREF none).
+     * Bugs: This function is currently not `nothrow` or `@nogc`.
+     * $(LREF Option.front) can be used as a workaround when these attributes
+     * are desired. See $(BUGZILLA 12647).
+     */
+    // this function is not nothrow due to bug 12647, and not @nogc.
+    // onAssertError isn't @nogc either.
+    ref inout(T) assumeNonEmpty() inout @property pure @safe
+    {
+        import core.exception : AssertError;
+        import std.exception : enforce;
+        enforce!AssertError(isDefined, T.stringof ~ " asssumption failed to hold");
```

Comment:
typo: assumption

We could just use `getOrElse(T.init)` instead of this. Alternatively we could throw an `Exception` so the user can deliberately catch it and handle it.